dali/kernels/imgproc/resample/resampling_setup.cc

```diff
@@ -98,80 +100,196 @@ SeparableResamplingSetup<spatial_ndim>::ComputeScaleAndROI(
-  int64_t compute_hv = size_horz * filter_support.x + out_area * filter_support.y;
-
-  // ...maybe fine tune the size/compute weights?
-  const float size_weight = 3;
-  float cost_vert = size_weight*size_vert + compute_vh;
-  float cost_horz = size_weight*size_horz + compute_hv;
-
-  int tmp_H, tmp_W;
-  if (cost_vert < cost_horz) {
-    desc.order = VertHorz();
-    tmp_H = out_H;
-    tmp_W = roi.extent().x;
-    desc.block_count[0] = div_ceil(out_H, block_size.y);
-    desc.block_count[1] = div_ceil(out_W, block_size.x);
-  } else {
-    desc.order = HorzVert();
-    tmp_H = roi.extent().y;
-    tmp_W = out_W;
-    desc.block_count[0] = div_ceil(out_W, block_size.x);
-    desc.block_count[1] = div_ceil(out_H, block_size.y);
+  ivec<spatial_ndim> filter_support;
+  for (int i = 0, d = spatial_ndim - 1; i < spatial_ndim; i++, d--) {
+    int support = desc.filter[d].support();
+    // NN filter has support -1, so we need the max() below
+    filter_support[i] = std::max(1, support);
   }
-  desc.tmp_shape(0) = { tmp_W, tmp_H };
 
-  for (int stage = 0; stage < 3; stage++) {
-    desc.strides[stage][0] = desc.shapes[stage].x * C;
+  desc.order = GetProcessingOrder(roi.extent(), out_size, filter_support);
+
+  {
+    ivec<spatial_ndim> pass_size = roi.extent();
+    for (int pass = 0; pass < spatial_ndim; pass++) {
+      int axis = desc.order[pass];
+
+      if (pass < spatial_ndim - 1) {
+        // for the last pass, the shape is out_size and is already set
+        pass_size[axis] = out_size[axis];
+        desc.tmp_shape(pass) = pass_size;
+      }
+    }
+  }
+
+  ComputeBlockLayout(desc);
+
+  // this sets strides and offsets for all stages, including input and output
+  for (int stage = 0; stage <= spatial_ndim; stage++) {
+    ptrdiff_t stride = desc.channels;
+    // There're fewer strides than dimensions.
+    for (int d = 0; d < spatial_ndim - 1; d++) {
+      stride *= desc.shapes[stage][d];
+      desc.strides[stage][d] = stride;
+    }
     desc.offsets[stage] = 0;
   }
-  desc.channels = C;
 
-  if (desc.order == VertHorz()) {
-    desc.origin.x -= roi.lo.x;
-    desc.in_offset() += roi.lo.x * desc.channels;
-    desc.in_shape().x = roi.extent().x;
-  } else {
-    desc.origin.y -= roi.lo.y;
-    desc.in_offset() += roi.lo.y * desc.strides[0][0];
-    desc.in_shape().y = roi.extent().y;
+  {
+    int first_pass_axis = desc.order[0];
+    auto strides = cat<ptrdiff_t>(channels, desc.strides[0]);
+    for (int a = 0; a < spatial_ndim; a++) {
+      if (a != first_pass_axis) {
+        desc.origin[a] -= roi.lo[a];
+        desc.in_offset() += roi.lo[a] * strides[a];
+        desc.in_shape()[a] = roi.extent()[a];
+      }
+    }
   }
 }
 
-template <>
-void BatchResamplingSetup<2>::SetupBatch(
-    const TensorListShape<3> &in, const Params &params) {
-  if (!filters)
-    Initialize();
+template <int spatial_ndim>
+void BatchResamplingSetup<spatial_ndim>::SetupBatch(
```

Comment:
Can you describe a bit a high level idea behind functions in this module. It is hard to digest the details not knowing the overall idea behind this code.

Reply:
Maybe.... I mean - if I can remember the details after all this time ;)

Reply:
> if I can remember the details after all this time ;)

Then it is another reason to add it.

Reply:
Done.